File: DataStructure/JavaScript/QueueImp/Queue.js
'use strict';

const Node = require('./Node');

class Queue {
  constructor() {
    this.first = null;
  }

  enqueue(data) {
    const newNode = new Node(data);
    if (!this.first) {
      this.first = newNode;
      return this;
    }
    const current = this.first;
    while (current.next) {
      current = current.next;
    }
    current.next = newNode;
    return this;
  }

  dequeue() {
    const removed = this.first.data;
    this.first = this.first.next;
    return removed;
  }

  peek() {
    return this.first.data;
  }

  display() {
    let output = '';
    let current = this.first;
    while (current) {
      output += `${current.data} --> `;
      current = current.next;
    }
    return output;
  }
  isEmpty() {
    return this.first ? false : true;
  }

  length() {
    let counter = 0;
    let current = this.first;
    while (current) {
      counter++;
      current = current.next;
    }
    return counter;
  }
}

File: DataStructure/Java/QueueImp/QueueImpLL/Queue.java
package QueueImp.QueueImpLL;

public class Queue<T> {
    private Node<T> head;
    private Node<T> tail;

    void enqueue(T data){
        Node<T> newNode = new Node<>(data);
        if(head == null){
            head = tail = newNode;
            return;
        }
        tail.next = newNode;
        tail = newNode;
    }

    T dequeue(){
     T removed = head.data;
     head = head.next;
     if(head == null) {
         tail = null;
     }
     return removed;
    }

    T peek(){
        return head.data;
    }

    boolean isEmpty(){
        return head == null;
    }

    int length(){
        int counter = 0;
        Node<T> current = head;
        while (current != null){
            counter++;
            current = current.next;
        }
        return counter;
    }

    String display(){
        if(head == null){
            return "Empty";
        }
        StringBuilder sb = new StringBuilder();
        Node<T> current = head;
        while (current != null){
            sb.append(current.data).append(" --> ");
            current = current.next;
        }
        return sb.append("Null").toString();
    }
}
